operator/iamkeycontroller/create.go

```diff
@@ -54,45 +64,110 @@ func (p *IAMKeyPipeline) Create(ctx context.Context, mg resource.Managed) (manag
 
 // createIAMKey creates a new IAMKey in the project associated with the API Key and Secret.
 func (p *IAMKeyPipeline) createIAMKey(ctx *pipelineContext) error {
+
 	iamKey := ctx.iamKey
 	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("starting creation")
+
+	policyAllow := exooapi.IamServicePolicyRuleActionAllow
+	policyDeny := exooapi.IamServicePolicyRuleActionDeny
+	policyRules := exooapi.IamServicePolicyTypeRules
+
+	log.Info("IAM Role doesnt exists, creating", "keyName", ctx.iamKey.Spec.ForProvider.KeyName)
+	autogeneratedAppcatRole := &exooapi.IamRole{
+		Name:        &iamKey.Spec.ForProvider.KeyName,
+		Description: pointer.String("IAM Role for SOS+IAM creation, it was autogenerated by provider-exoscale"),
+		Permissions: &[]exooapi.IamRolePermissions{
+			exooapi.IamRolePermissionsBypassGovernanceRetention,
+		},
+		Editable: pointer.Bool(true),
+		Policy: &exooapi.IamPolicy{
+			DefaultServiceStrategy: exooapi.IamPolicyDefaultServiceStrategyDeny,
+			Services: exooapi.IamPolicy_Services{
+				AdditionalProperties: map[string]exooapi.IamServicePolicy{
+					"sos": {
+						Type: &policyRules,
+						Rules: &[]exooapi.IamServicePolicyRule{
+							{
+								Action: &policyDeny,
+								// bucket name in this rule must be in single quotes, otherwise it doesn't work and user has access to everything
+								Expression: pointer.String("resources.bucket != " + "'" + strings.Split(iamKey.Spec.ForProvider.KeyName, ".")[1] + "'"),
+							},
+							{
+								Action:     &policyAllow,
+								Expression: pointer.String("true"),
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// send request
+	resp, err := ExecuteRequest(ctx, "POST", ctx.iamKey.Spec.ForProvider.Zone, "/v2/iam-role", p.apiKey, p.apiSecret, autogeneratedAppcatRole)
+	if err != nil {
+		return err
+	}
+
+	iamRole := exooapi.Operation{}
+
+	// unmarshall response to iamRoles
+	err = json.NewDecoder(resp.Body).Decode(&iamRole)
+	if err != nil {
+		log.Error(err, "Cannot unmarshall response to iamRolesResponse")
+		return err
+	}
+
+	iamRoleID := *iamRole.Reference.Id
+	log.Info("IAM Role created", "iamRoleID", iamRoleID)
+
+	log.Info("IAM Key doesnt exists, creating", "keyName", ctx.iamKey.Spec.ForProvider.KeyName)
+	newIamKey := exooapi.IamApiKey{
+		RoleId: &iamRoleID,
+		Name:   &ctx.iamKey.Spec.ForProvider.KeyName,
+	}
+
+	b1, _ := json.MarshalIndent(newIamKey, "\t", "\t")
+	fmt.Println("\n\n\n\n\n\n\n\n\n\n", string(b1))
 
-	var keyResources []exoscalesdk.IAMAccessKeyResource
-	for _, bucket := range iamKey.Spec.ForProvider.Services.SOS.Buckets {
-		keyResource := exoscalesdk.IAMAccessKeyResource{
-			Domain:       SOSResourceDomain,
-			ResourceName: bucket,
-			ResourceType: BucketResourceType,
+	b1, _ = json.MarshalIndent(iamRole, "\t", "\t")
+	fmt.Println("\n\n\n\n\n\n\n\n\n\n", string(b1))
+
+	// since their API is async and it needs a moment to create the IAM Role, we need to wait for it
+	for i := 0; i < 10; i++ {
```

Comment:
Nitpick: This is ugly. We should implement a [RoundTripper](https://morioh.com/p/5beb43562e87) for the client and abstract this loop

Reply:
Ok, I've familiarized with RoundTripper and I think it's great tool but for bigger use cases, it does basically the same thing but in different place. I think that implementation of round tripper here, would just make code more complex. If You insist then no problem, but I can't see value for single request

Reply:
Agreed, if they start to switch everything to async some abstraction for this would be nice... (if they don't provide it in egoscale)